### PR TITLE
Add CAST(varchar as decimal)

### PR DESCRIPTION
**Pull request by jinchengchenghh**
Spark implementation: https://github.com/apache/spark/blob/branch-3.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L822

Derives from Arrow function `DecimalFromString`.
Arrow implementation: https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/decimal.cc#L637

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
Could you help take a look? @Yuhta

**Review by mbasmanova — COMMENTED** [review]
@majetideepak Deepak, would you help review this PR?

**Comment by rui-mo** [comment]
Can we add Spark's implementation to the PR description, and ensure all cases are considered in unit test?

**Review by majetideepak — COMMENTED** [review]
@jinchengchenghh Is this emulating the Java implementation? I see a lot of string manipulations.
Can we avoid them and work with offsets for the dot and exponent on the original string?

**Comment by jinchengchenghh** [comment]
No, java uses JDK API to convert to decimal.
```
  private def stringToJavaBigDecimal(str: UTF8String): JavaBigDecimal = {
    // According the benchmark test,  `s.toString.trim` is much faster than `s.trim.toString`.
    // Please refer to https://github.com/apache/spark/pull/26640
    new JavaBigDecimal(str.toString.trim)
  }
```
Where does the convert number with exponent come from? @rui-mo

**Comment by rui-mo** [comment]
> Where does the convert number with exponent come from? @rui-mo

@jinchengchenghh They are from Spark unit tests.

**Comment by jinchengchenghh** [comment]
> @jinchengchenghh Is this emulating the Java implementation? I see a lot of string manipulations. Can we avoid them and work with offsets for the dot and exponent on the original string?

Convert number with exponent comes from @rui-mo , I will try to optimize it.

I found postgresql has implementation to convert exponent varchar to numeric, https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c#L4515, I'm not sure if we can write as this.
Or we should implement by ourselves?

**Comment by majetideepak** [comment]
@jinchengchenghh This seems to be the main implementation https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c#L6942
This is very C-style ish. I don't think we need to be this minimal but is a good start. I don't have an answer here.
Maybe you look at other open-source implementations like Impala or DuckDB as well.

**Comment by jinchengchenghh** [comment]
Update to Arrow implementation, can you help to review again? Thanks! @majetideepak

File: velox/expression/CastExpr-inl.h [hunk 1]
```diff
@@ -206,6 +206,30 @@ void CastExpr::applyIntToDecimalCastKernel(
       });
 }
 
+template <typename TOutput>
+void applyVarCharToDecimalCastKernel(
```
Comment (majetideepak) [inline velox/expression/CastExpr-inl.h:210]: nit: `VarChar` -> `Varchar`

File: velox/expression/CastExpr-inl.h [hunk 2]
```diff
@@ -206,6 +206,30 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr castResult) {
```
Comment (majetideepak) [inline velox/expression/CastExpr-inl.h:215]: `VectorPtr& castResult`

File: velox/expression/tests/CastExprTest.cpp [hunk 3]
```diff
@@ -1470,6 +1470,100 @@ TEST_F(CastExprTest, integerToDecimal) {
...
+// The result is obtained by select cast('31.4e-2' as decimal(12, 2)).
+TEST_F(CastExprTest, varcharToDecimal) {
+  auto input = makeFlatVector<StringView>(
+      {"9999999999.99",
+       "15",
+       "1.5",
+       "-1.5",
+       "1.556",
+       "1.554",
+       ("1.556" + std::string(32, '1')).data(),
```
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:1482]: include `("1.556" + std::string(32, '9')).data()` as well?

File: velox/type/DecimalUtil.h [hunk 4]
```diff
@@ -272,5 +329,87 @@ class DecimalUtil {
...
+      out->exponent =
+          folly::to<int32_t>(folly::StringPiece(s + pos, size - pos));
+      return true;
+    }
+    return pos == size;
+  }
+
+  /// Multiple out by the appropriate power of 10 necessary to add source parsed
+  /// as int128_t and then adds the parsed value of source.
+  inline static void shiftAndAdd(std::string_view input, int128_t& out) {
```
Comment (majetideepak) [inline velox/type/DecimalUtil.h:404]: nit: use int128_t in and return int128_t out?

File: velox/expression/tests/CastExprTest.cpp [hunk 5]
```diff
@@ -1470,6 +1470,100 @@ TEST_F(CastExprTest, integerToDecimal) {
...
+
+  std::vector<std::string> illegalInputs = {
+      std::string(280, '9'), "0.0444a", ""};
+  for (auto& input : illegalInputs) {
+    VELOX_ASSERT_THROW(
+        testComplexCast(
+            "c0",
+            makeConstant<StringView>(input.data(), 1),
+            makeConstant<int128_t>(1, 1, DECIMAL(38, 0))),
+        "")
```
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:1563]: Can you unroll the loop and include the Error message for each input?

File: velox/expression/CastExpr-inl.h [hunk 6]
```diff
@@ -206,6 +206,30 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+  auto sourceVector = input.as<SimpleVector<StringView>>();
+  auto castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+  context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
```
Comment (majetideepak) [inline velox/expression/CastExpr-inl.h:220]: Should this be `applyToSelectedNoThrowLocal` like others?

File: velox/type/DecimalUtil.h (hunk 4 above)
Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:404]: Yes, out = out * input().size + valueOf(input)

File: velox/expression/CastExpr-inl.h (hunk 6 above)
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h]: Fixed all, thanks!

**Comment by jinchengchenghh** [comment]
Do you have further comments? @majetideepak

**Review by majetideepak — APPROVED** [review]
Thanks, @jinchengchenghh

**Comment by jinchengchenghh** [comment]
Could you help merge it? Thanks! @mbasmanova

**Review by mbasmanova — COMMENTED** [review]
@majetideepak @jinchengchenghh Have we verified that this behavior matches Presto?

**Comment by jinchengchenghh** [comment]
I have verified it in Presto. @mbasmanova

**Comment by jinchengchenghh** [comment]
Presto throws exception when input is illegal while Spark return NULL, this behavior also matches Presto.

**Comment by mbasmanova** [comment]
> I have verified it in Presto.

Thanks. Would you update PR description to clarify?

> Presto throws exception when input is illegal while Spark return NULL, this behavior also matches Presto.

Sounds like there are differences in behavior for Presto and Spark. Is this so? Should there be some fork in the code then?

File: velox/expression/CastExpr-inl.h [hunk 7]
```diff
@@ -206,6 +206,31 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+void CastExpr::applyVarcharToDecimalCastKernel(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<StringView>>();
+  auto castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:219]: getDecimalPrecisionScale returns a temp value, hence, it should be captured by value, not by reference.

File: velox/type/DecimalUtil.h [hunk 8]
```diff
@@ -147,6 +147,63 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  // Derives from Arrow function DecimalFromString.
+  template <typename TOutput>
+  inline static std::optional<TOutput> rescaleVarchar(
+      const StringView& s,
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:153]: StringView should be passed by value.

File: velox/type/DecimalUtil.h [hunk 9]
```diff
@@ -147,6 +147,63 @@ class DecimalUtil {
...
+      const int toPrecision,
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:154]: drop 'const' here and on the next line

**Comment by jinchengchenghh** [comment]
> > I have verified it in Presto.
> 
> Thanks. Would you update PR description to clarify?
> 
> > Presto throws exception when input is illegal while Spark return NULL, this behavior also matches Presto.
> 
> Sounds like there are differences in behavior for Presto and Spark. Is this so? Should there be some fork in the code then?

This is the difference for all the Cast behavior, we may need to consider the strategy to treat this in the future.

File: velox/type/DecimalUtil.h [hunk 10]
```diff
@@ -147,6 +147,63 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  // Derives from Arrow function DecimalFromString.
+  template <typename TOutput>
+  inline static std::optional<TOutput> rescaleVarchar(
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:152]: It would be nice to extract this change into a separate PR with its own tests. 

Does all this code need to be in the .h file? Can some of it be moved into .cpp file to speed up builds?

File: velox/type/DecimalUtil.h [hunk 11]
```diff
@@ -272,5 +329,87 @@ class DecimalUtil {
...
+    }
+    if (out->wholeDigits.empty() && out->fractionalDigits.empty()) {
+      // Need at least some digits (whole or fractional).
+      return false;
+    }
+    if (pos == size) {
+      return true;
+    }
+    // Optional exponent.
+    if (s[pos] == 'e' || s[pos] == 'E') {
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:388]: It would be nice to update the documentation to explain valid formats of the strings that can be casted to decimals.

**Comment by mbasmanova** [comment]
> This is the difference for all the Cast behavior, we may need to consider the strategy to treat this in the future.

Does Spark ever throw from CAST? Would it be correct to say that Spark's CAST behavior matches Presto's TRY_CAST?

It is important to mention this in the PR description. Right now, the PR description suggest that the implementation matches Spark, which seems to be not the case.

File: velox/type/DecimalUtil.h (hunk 10 above)
Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:152]: This one is only used in Cast, does it need to be in DecimalUtil.h? Can I move it to `velox/expression/CastExpr-inl.h` and add `velox/expression/CastExpr-inl.cpp`?

I just align with DecimalUtil::rescaleInt.

**Comment by jinchengchenghh** [comment]
> https://github.com/apache/spark/blob/branch-3.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L822

Spark has ANSI mode, default true means return NULL.
https://github.com/apache/spark/blob/branch-3.3/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L2840

**Comment by mbasmanova** [comment]
> Spark has ANSI mode, default true means return NULL, false means throw exception.

I see. Does Gluten support ANSI mode? If so, this implementation seems incorrect, no?

File: velox/type/DecimalUtil.h (hunk 10 above)
Comment (mbasmanova) [reply velox/type/DecimalUtil.h:152]: I'll let @majetideepak to comment on that.

**Comment by jinchengchenghh** [comment]
Gluten only supports ANSI false mode, in another words, return NULL.
Yes, we may need to add a config to support part of ANSI mode for all the cast behavior.

**Comment by mbasmanova** [comment]
> Gluten only supports ANSI false mode, in another words, return NULL.

Got it. Curious, what's the reason for not supporting ANSI mode in Gluten?

> Yes, we may need to add a config to support part of ANSI mode for all the cast behavior.

It would be nice to figure out how to do that sooner rather than later. CC: @rui-mo

**Comment by jinchengchenghh** [comment]
I don't know some background before, I asked @PHILO-HE. 
Velox has nullOnFailure mode which matches Spark ANSI false mode. So this implementation is correct. @mbasmanova

File: velox/type/DecimalUtil.h [hunk 12]
```diff
@@ -272,5 +327,87 @@ class DecimalUtil {
...
+      ++pos;
+      if (pos != size && s[pos] == '+') {
+        ++pos;
+      }
+      out->hasExponent = true;
+      folly::StringPiece p = {s + pos, size - pos};
+      out->exponent =
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:393]: Why do you use folly::to here and not parseDigitsRun?

What happens if folly::to throws? Do we get the "right" error message?

```
> SELECT cast('1.2e-5d' as decimal(12, 10));
Query 20230822_130057_10377_gk7h6 failed: Cannot cast VARCHAR '1.2e-5d' to DECIMAL(12, 10). Value is not a number.
```

File: velox/type/DecimalUtil.h [hunk 13]
```diff
@@ -272,5 +327,87 @@ class DecimalUtil {
...
+      // Need at least some digits (whole or fractional).
+      return false;
+    }
+    if (pos == size) {
+      return true;
+    }
+    // Optional exponent.
+    if (s[pos] == 'e' || s[pos] == 'E') {
+      ++pos;
+      if (pos != size && s[pos] == '+') {
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:388]: Should we also parse  '1.2e-5d'?

**Review by mbasmanova — COMMENTED** [review]
Thank you for updating the documentation. It is much clearer now. Some further comments about the implementation.

**Comment by rui-mo** [comment]
> It would be nice to figure out how to do that sooner rather than later.

@mbasmanova For now, if Gluten detects the ANSI config is ON, it fallbacks all the computing to vanilla Spark. In Spark, the ANSI config controls a lot of cases to throw exception or return null, and I'm wondering if a general config like `throw_or_return_null` can be added in Velox. It could be used to control cast behavior, agg functions, arithmetic functions etc. How do you think?

File: velox/expression/CastExpr-inl.h [hunk 14]
```diff
@@ -206,6 +206,31 @@ void CastExpr::applyIntToDecimalCastKernel(
       });
 }
 
+template <typename TOutput>
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:209]: Just align with https://github.com/facebookincubator/velox/blob/main/velox/expression/CastExpr-inl.h#L163 and https://github.com/facebookincubator/velox/blob/main/velox/expression/CastExpr-inl.h#L184

Do we need to change to the new one `T`?

File: velox/expression/CastExpr-inl.h [hunk 15]
```diff
@@ -206,6 +206,31 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+void CastExpr::applyVarcharToDecimalCastKernel(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<StringView>>();
+  auto castResultRawBuffer =
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:217]: same as above

**Comment by mbasmanova** [comment]
> > It would be nice to figure out how to do that sooner rather than later.
> 
> @mbasmanova For now, if Gluten detects the ANSI config is ON, it fallbacks all the computing to vanilla Spark. In Spark, the ANSI config controls a lot of cases to throw exception or return null, and I'm wondering if a general config like `throw_or_return_null` can be added in Velox. It could be used to control cast behavior, agg functions, arithmetic functions etc. How do you think?

Some high-level thoughts.

Throw-or-null logic is tricky to implement. Exceptions are expensive, hence, catching them and returning null is not ideal. We have seen how this approach resulted in major performance problems in TRY_CAST. See https://github.com/facebookincubator/velox/pull/5801 CC @laithsakka 

To make this performant, the code needs to not use exceptions and rather return error-or-result. This is inconvenient for developers and requires major refactoring in Velox. See https://abseil.io/docs/cpp/guides/status

File: velox/type/DecimalUtil.h [hunk 16]
```diff
@@ -147,6 +147,61 @@ class DecimalUtil {
...
+  inline static std::optional<TOutput>
+  rescaleVarchar(const StringView s, int toPrecision, int toScale) {
+    DecimalComponents dec;
+    VELOX_USER_CHECK(
+        parseDecimalComponents(s.data(), s.size(), &dec),
+        "Value {} is not a number",
+        s);
+
+    // Count number of significant digits (without leading zeros).
+    size_t firstNonZero = dec.wholeDigits.find_first_not_of('0');
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:161]: I have added the test `+09`

File: velox/type/DecimalUtil.h [hunk 17]
```diff
@@ -272,5 +327,87 @@ class DecimalUtil {
...
+  struct DecimalComponents {
+    std::string_view wholeDigits;
+    std::string_view fractionalDigits;
+    int32_t exponent = 0;
+    char sign = 0;
+    bool hasExponent = false;
+  };
+
+  inline static size_t parseDigitsRun(
+      const char* s,
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:341]: Because we need to construct folly::StringPiece as `folly::StringPiece p = {s + pos, size - pos};`

File: velox/type/DecimalUtil.h (hunk 12 above)
Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:393]: parseDigitsRun just return the exponent string, here we need the integer exponent

File: velox/expression/CastExpr-inl.h [hunk 18]
```diff
@@ -30,10 +30,10 @@ inline std::string makeErrorMessage(
     const TypePtr& toType,
     const std::string& details = "") {
   return fmt::format(
-      "Failed to cast from {} to {}: {}. {}",
+      "Cannot cast {} '{}' to {}. {}",
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr-inl.h:33]: I find the cast fail message is different with Presto, do I need to change it in a separate PR or use the old error message? @mbasmanova

File: velox/type/Type.h [hunk 19]
```diff
@@ -662,7 +663,7 @@ class DecimalType : public ScalarType<KIND> {
   }
 
   std::string toString() const override {
-    return fmt::format("DECIMAL({},{})", precision(), scale());
+    return fmt::format("DECIMAL({}, {})", precision(), scale());
```
Comment (jinchengchenghh) [inline velox/type/Type.h:666]: Different with Presto, separate PR, this PR or not change it? @mbasmanova

File: velox/expression/CastExpr-inl.h (hunk 18 above)
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:33]: Same PR should be fine. Thank you for noticing and fixing.

File: velox/type/DecimalUtil.h (hunk 10 above)
Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h]: Hi, @majetideepak Can you take a time to look into this discussion?

**Comment by rui-mo** [comment]
> To make this performant, the code needs to not use exceptions and rather return error-or-result. This is inconvenient for developers and requires major refactoring in Velox. See https://abseil.io/docs/cpp/guides/status

@mbasmanova Thanks for the details. According to above, maybe there are two options, right?

1. Use a flag `bool nullOnFailure` just like #5801
2. Use absl::Status (maybe after Velox's refactor)

**Comment by mbasmanova** [comment]
@rui-mo 

> Thanks for the details. According to above, maybe there are two options, right?
> 
> 1. Use a flag `bool nullOnFailure` just like [Fix and optimize try_cast #5801](https://github.com/facebookincubator/velox/pull/5801)
> 2. Use absl::Status (maybe after Velox's refactor)

That's right. At this point, (1) is probably the only option.

File: velox/type/DecimalUtil.h (hunk 10 above)
Comment (majetideepak) [reply velox/type/DecimalUtil.h]: @jinchengchenghh I am okay with moving decimal cast helper functions to CastExpr* if they are only used there. But since these are template functions, you must put them in a header file right?
Am I missing something?

Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h]: We can put it to cpp file https://stackoverflow.com/questions/115703/storing-c-template-function-definitions-in-a-cpp-file
But put it to .h file maybe more simple, so I will put it in CastExpr-inl.h

**Comment by jinchengchenghh** [comment]
Address all the comments, do you have further comments? @mbasmanova

File: velox/type/Type.h (hunk 19 above)
Comment (mbasmanova) [inline velox/type/Type.h:666]: This change seems unrelated and causes lots of other unrelated changes making this PR larger than needed. It looks like I misunderstood your earlier question as I didn't realize the change will be wide spread. Can it be extracted into a separate PR?

File: velox/expression/CastExpr-inl.h [hunk 20]
```diff
@@ -46,8 +46,175 @@ inline std::exception_ptr makeBadCastException(
...
+    // Need at least some digits (whole or fractional).
+    return std::nullopt;
+  }
+  if (pos == size) {
+    return out;
+  }
+  // Optional exponent.
+  if (s[pos] == 'e' || s[pos] == 'E') {
+    ++pos;
+    if (pos != size && s[pos] == '+') {
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:120]: Is it not possible to have a '-' here?

File: velox/expression/CastExpr-inl.h [hunk 21]
```diff
@@ -46,8 +46,175 @@ inline std::exception_ptr makeBadCastException(
...
+}
+
+/// Multiple out by the appropriate power of 10 necessary to add source parsed
+/// as int128_t and then adds the parsed value of source.
+bool shiftAndAdd(std::string_view input, int128_t& out) {
+  auto length = input.size();
+  if (length == 0) {
+    return true;
+  }
+  out = checkedMultiply<int128_t>(out, DecimalUtil::kPowersOfTen[length]);
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:142]: Do checkedMultiply and checkedPlus throw exceptions? The rest of the code is being careful to not do that.

File: velox/expression/CastExpr-inl.h [hunk 22]
```diff
@@ -46,8 +46,175 @@ inline std::exception_ptr makeBadCastException(
...
+  auto decimalComponentsOpt = parseDecimalComponents(s.data(), s.size());
+  if (!decimalComponentsOpt.has_value()) {
+    error = "Value is not a number.";
+    return std::nullopt;
+  }
+  auto decimalComponents = decimalComponentsOpt.value();
+
+  // Count number of significant digits (without leading zeros).
+  size_t firstNonZero = decimalComponents.wholeDigits.find_first_not_of('0');
+  size_t significantDigits = decimalComponents.fractionalDigits.size();
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:175]: Should we ignore trailing zeros in fractionalDigits?

File: velox/expression/CastExpr-inl.h [hunk 23]
```diff
@@ -46,8 +46,175 @@ inline std::exception_ptr makeBadCastException(
...
+  if (firstNonZero != std::string::npos) {
+    significantDigits += decimalComponents.wholeDigits.size() - firstNonZero;
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:177]: Let's update the documentation to include examples with multiple zeros at the start and end: '0000.123' and '.12300000000'

File: velox/expression/CastExpr-inl.h (hunk 21 above)
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:142]: Yes, should we add function `multiply` and `plus` with argument `bool& overflow` to https://github.com/facebookincubator/velox/blob/main/velox/common/base/CheckedArithmetic.h ?

File: velox/expression/CastExpr-inl.h (hunk 22 above)
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:175]: Spark and Presto has different result for that, how can we handle this case?
```
presto:default> select cast('1. 23 ' as decimal(38, 0));
Query 20230828_155901_00012_sqyi8 failed: Cannot cast VARCHAR '1. 23 ' to DECIMAL(38, 0). Value is not a number.

spark-sql> select cast('1.23 ' as decimal(38, 0));
1

```

File: velox/expression/CastExpr-inl.h (hunk 20 above)
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:120]: Here we convert `+10` to `10`, then convert it to int32 by folly, convert `-10` by folly too.

File: velox/expression/CastExpr-inl.h (hunk 22 above)
Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:175]: Here we will check the trailing space, https://github.com/facebookincubator/velox/pull/5307
So the behavior is same with Presto.

Comment (jinchengchenghh) [reply velox/expression/CastExpr-inl.h:175]: Can we align with Presto until this issue https://github.com/facebookincubator/velox/issues/4876 resolved?

File: velox/expression/CastExpr-inl.h (hunk 21 above)
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:142]: Maybe we could use __builtin_add_overflow and __builtin_add_overflow directly here.

**Comment by jinchengchenghh** [comment]
pending by https://github.com/facebookincubator/velox/discussions/6323

File: velox/expression/CastExpr-inl.h (hunk 22 above)
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h:260]: Opened https://github.com/facebookincubator/velox/pull/7377.

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 24]
```diff
@@ -236,7 +236,7 @@ class CastBaseTest : public FunctionBaseTest {
     testCast(fromType, toType, inputVector, expectedVector);
   }
 
-  template <typename TFrom, typename TTo>
+  template <typename TFrom>
```
Comment (rui-mo) [inline velox/functions/prestosql/tests/CastBaseTest.h:239]: Remove `typename TTo` as it is not in use. Do I need to do this refactor in a separate PR?

**Comment by rui-mo** [comment]
Hi @mbasmanova @majetideepak This PR has been updated. Could you help review again? Changes including 1) using Status instead of throwing exception 2) using cast hook to support different behaviors on white-spaces handling between Presto and Spark.

File: velox/type/DecimalUtil.h [hunk 25]
```diff
@@ -146,40 +147,40 @@ class DecimalUtil {
   }
 
   template <typename TInput, typename TOutput>
-  inline static std::optional<TOutput> rescaleWithRoundUp(
+  inline static Status rescaleWithRoundUp(
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:150]: This is PR is a bit large and complicated. Would it be possible to extract this particular change into a separate PR to make review easier?

File: velox/functions/prestosql/tests/CastBaseTest.h (hunk 24 above)
Comment (mbasmanova) [reply velox/functions/prestosql/tests/CastBaseTest.h:239]: This is a nice change. It would be great to extract it into a separate PR to make reviewing easier and faster. Thanks.

File: velox/expression/tests/CastExprTest.cpp [hunk 26]
```diff
@@ -1692,6 +1706,259 @@ TEST_F(CastExprTest, boolToDecimal) {
...
+      "c0",
+      makeFlatVector<StringView>(
+          {StringView(fractionRoundDown), StringView(fractionRoundDownExp)}),
+      makeConstant<int128_t>(DecimalUtil::kLongDecimalMax, 2, DECIMAL(38, 38)));
+
+  // Function shiftAndAdd overflows when shifting whole digits.
+  testThrow<StringView>(
+      VARCHAR(),
+      DECIMAL(38, 0),
+      toStringViews({std::string(280, '9')}),
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/tests/CastExprTest.cpp:1847]: Notice that StringView doesn't own the string, hence, it is invalid to create StringView from a temporary std::string.

**Review by mbasmanova — COMMENTED** [review]
@rui-mo Rui, thank you for working on this change. I'll try to review it early next week. Please, ping me if you don't hear from me before Tue.

File: velox/functions/prestosql/tests/CastBaseTest.h (hunk 24 above)
Comment (rui-mo) [reply velox/functions/prestosql/tests/CastBaseTest.h:239]: Extract this change to https://github.com/facebookincubator/velox/pull/8464. Thanks.

File: velox/type/DecimalUtil.h (hunk 25 above)
Comment (rui-mo) [reply velox/type/DecimalUtil.h:150]: Extract this change to https://github.com/facebookincubator/velox/pull/8465.

File: velox/expression/tests/CastExprTest.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/tests/CastExprTest.cpp]: Thanks for the comment. Changed to use `testThrow<std::string>`.

**Review by mbasmanova — APPROVED** [review]
@rui-mo Looks good to me % small comments.

File: velox/docs/functions/spark/conversion.rst [hunk 27]
```diff
@@ -162,3 +162,23 @@ Invalid examples
...
   SELECT cast('2012.10.23' as date); -- Invalid argument
+
+Cast to Decimal
+---------------
+
+From varchar
+^^^^^^^^^^^^
+
+Casting varchar to a decimal of given precision and scale is allowed.
+Leading and trailing white-spaces in input varchars are allowed.
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/docs/functions/spark/conversion.rst]: That's the case, so I assume we don't need add duplicate description. Clarify in the description.

File: velox/expression/CastExpr-inl.h [hunk 28]
```diff
@@ -52,6 +52,22 @@ inline std::exception_ptr makeBadCastException(
...
+/// Represent the varchar fragment.
+///
+/// For example:
+/// | value | wholeDigits | fractionalDigits | exponent | sign |
+/// | 9999999999.99 | 9999999999 | 99 | nullopt | 1 |
+/// | 15 | 15 |  | nullopt | 1 |
+/// | 1.5 | 1 | 5 | nullopt | 1 |
+/// | -1.5 | 1 | 5 | nullopt | -1 |
+/// | 31.523e-2 | 31 | 523 | -2 | 1 |
+struct DecimalComponents {
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Moved them to `detail` namespace. Thanks.

File: velox/expression/CastExpr-inl.h [hunk 29]
```diff
@@ -112,6 +128,212 @@ StringView convertToStringView(
...
+// Extract a string view of continuous digits.
+size_t
+extractDigits(const char* s, size_t start, size_t size, std::string_view& out) {
+  size_t pos = start;
+  for (; pos < size; ++pos) {
+    if (!std::isdigit(s[pos])) {
+      break;
+    }
+  }
+  out = std::string_view(s + start, pos - start);
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Updated, thanks.

File: velox/expression/CastExpr-inl.h [hunk 30]
```diff
@@ -112,6 +128,212 @@ StringView convertToStringView(
...
+  }
+  // Optional exponent.
+  if (s[pos] == 'e' || s[pos] == 'E') {
+    ++pos;
+    bool withSign = pos < size && (s[pos] == '+' || s[pos] == '-');
+    if (withSign && pos == size - 1) {
+      // The exponent part only contains sign.
+      return std::nullopt;
+    }
+    // Make sure all chars after sign are digits, as folly does not prevent
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Changed to use folly::to here.

File: velox/expression/CastExpr-inl.h [hunk 31]
```diff
@@ -112,6 +128,212 @@ StringView convertToStringView(
...
+    // leading and trailing whitespaces.
+    for (auto i = (size_t)withSign; i < size - pos; ++i) {
+      if (!std::isdigit(s[pos + i])) {
+        return std::nullopt;
+      }
+    }
+    const auto tryExp =
+        folly::tryTo<int32_t>(folly::StringPiece(s + pos, size - pos));
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Changed to use folly::to here.

File: velox/expression/CastExpr-inl.h [hunk 32]
```diff
@@ -112,6 +128,212 @@ StringView convertToStringView(
...
+/// the outcome of computing.
+template <typename T>
+Status toDecimalValue(
+    const StringView s,
+    int toPrecision,
+    int toScale,
+    T& decimalValue) {
+  const auto decimalComponentsOpt = parseDecimalComponents(s.data(), s.size());
+  if (!decimalComponentsOpt.has_value()) {
+    return Status::UserError("Value is not a number.");
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h:165]: The full error message is like below, which contains the value. Do we need to add value here? 
> Cannot cast VARCHAR '1.23 ' to DECIMAL(38, 0). Value is not a number.

File: velox/expression/CastExpr-inl.h [hunk 33]
```diff
@@ -112,6 +128,212 @@ StringView convertToStringView(
...
+    VELOX_DCHECK(!tryExp.hasError());
+    out.exponent = tryExp.value();
+    return out;
+  }
+  return pos == size ? std::optional<DecimalComponents>(out) : std::nullopt;
+}
+
+/// Multiples the output by required power of 10, and adds the parsed value of
+/// input with the rescaled ouput.
+bool shiftAndAdd(std::string_view input, int128_t& out) {
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Thanks for the pointer. I removed this function, and combined the two calls as `parseHugeInt`. The new function creates a huge int from decimal components.

File: velox/expression/CastExpr-inl.h (hunk 32 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:165]: Got it. Thank you for clarifying. Looks good.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:165]: Actually, how would a user understand why 1.23 is not a number?

[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h:165]: For this case, it is because of the trailing white-space. Do we need to add the detailed reason?

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:165]: It would be nice to make error message as clear as possible to avoid unnecessary support calls.

[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Thanks. Updated. The error message now contains status message from parseDecimalComponent.

**Comment by rui-mo** [comment]
@mbasmanova Above comments were fixed. Could you review again? Thanks!

**Review by Yuhta — APPROVED** [review]

**Comment by mbasmanova** [comment]
@rui-mo Would you rebase?

**Comment by rui-mo** [comment]
Thank you all for your great help!

**Comment by mbasmanova** [comment]
@rui-mo Thank you for the contribution. It was quite a lot of work. Much appreciated.

**Comment by jinchengchenghh** [comment]
Many thanks for your kind and warm help！@rui-mo
